pkg/security/api/v1/types.go

```diff
@@ -0,0 +1,86 @@
+	unversioned.TypeMeta `json:",inline"`
+
+	// Spec defines specification for the PodSecurityPolicySubjectReview.
+	Spec PodSecurityPolicySubjectReviewSpec `json:"spec"`
+
+	// Status represents the current information/status for the PodSecurityPolicySubjectReview.
+	Status PodSecurityPolicySubjectReviewStatus `json:"status,omitempty"`
+}
+
+// PodSecurityPolicySubjectReviewSpec defines specification for PodSecurityPolicySubjectReview
+type PodSecurityPolicySubjectReviewSpec struct {
+	// PodSpec is the PodSpec to check.
+	PodSpec kapi.PodSpec `json:"podSpec"`
+
+	// User is the user you're testing for.
+	// If you specify "User" but not "Group", then is it interpreted as "What if User were not a member of any groups.
+	// If User and Groups are empty, then the check is performed using *only* the ServiceAccountName in the PodSpec.
+	User string `json:"user,omitempty"`
+
+	// Groups is the groups you're testing for.
+	Groups []string `json:"groups,omitempty"`
+}
+
+// PodSecurityPolicySubjectReviewStatus contains information/status for PodSecurityPolicySubjectReview.
+type PodSecurityPolicySubjectReviewStatus struct {
+	// AllowedBy is a reference to the rule that allows the PodSpec.  A `nil`, indicates that it was denied.
+	AllowedBy *kapi.ObjectReference `json:"allowedBy"`
+
+	// Reason is optional.  It indicates why a request was allowed or denied.
+	Reason string `json:"reason,omitempty"`
+
+	// PodSpec is the PodSpec after the defaulting is applied.
+	PodSpec kapi.PodSpec `json:"podSpec"`
+}
+
+// PodSecurityPolicySelfSubjectReview checks whether this user/SA tuple can create the PodSpec
+type PodSecurityPolicySelfSubjectReview struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Spec defines specification the PodSecurityPolicySelfSubjectReview.
+	Spec PodSecurityPolicySelfSubjectReviewSpec `json:"spec"`
+
+	// Status represents the current information/status for the PodSecurityPolicySelfSubjectReview.
+	Status PodSecurityPolicySubjectReviewStatus `json:"status,omitempty"`
+}
+
+// PodSecurityPolicySelfSubjectReviewSpec contains specification for PodSecurityPolicySelfSubjectReview.
+type PodSecurityPolicySelfSubjectReviewSpec struct {
+	// PodSpec is the PodSpec to check.
+	PodSpec kapi.PodSpec `json:"podSpec"`
+}
+
+// PodSecurityPolicyReview checks which service accounts (not users, since that would be cluster-wide) can create the `PodSpec` in question.
+type PodSecurityPolicyReview struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Spec is the PodSecurityPolicy to check.
+	Spec PodSecurityPolicyReviewSpec `json:"spec"`
+
+	// Status represents the current information/status for the PodSecurityPolicyReview.
+	Status PodSecurityPolicyReviewStatus `json:"status,omitempty"`
+}
+
+// PodSecurityPolicyReviewSpec defines specification for PodSecurityPolicyReview
+type PodSecurityPolicyReviewSpec struct {
+	// PodSpec is the PodSpec to check. The ServiceAccountName field is ignored for this check.
```

Comment:
Do I need a way to do that?  It doesn't seem common.  I would expect to be
able to do this:

sclient.Check(&PodSecurityPolicySubjectReview{Spec: Spec{PodSpec:
&pod.Spec}})

and have it do the right thing (use the SA defined in the pod, or default
it to "default").  I think specifying ServiceAccountNames is an optional
side path, and yes, if I specify SAN it should override what's in the pod
spec.

On Thu, May 26, 2016 at 12:36 PM, David Eads <notifications@github.com>
wrote:

> In pkg/security/api/v1/types.go
> <https://github.com/openshift/origin/pull/8941#discussion_r64778437>:
>
> > +}
> > +
> > +// PodSecurityPolicyReview checks which service accounts (not users, since that would be cluster-wide) can create the `PodSpec` in question.
> > +type PodSecurityPolicyReview struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +
> > +	// Spec is the PodSecurityPolicy to check.
> > +	Spec PodSecurityPolicyReviewSpec `json:"spec"`
> > +
> > +	// Status represents the current information/status for the PodSecurityPolicyReview.
> > +	Status PodSecurityPolicyReviewStatus `json:"status,omitempty"`
> > +}
> > +
> > +// PodSecurityPolicyReviewSpec defines specification for PodSecurityPolicyReview
> > +type PodSecurityPolicyReviewSpec struct {
> > +	// PodSpec is the PodSpec to check. The ServiceAccountName field is ignored for this check.
>
> Why is service account name ignored? It should be used unless
> ServiceAccountNames is checked.
>
> You need a way to indicate: check all SAs in this namespace, I don't know
> all of their names. To express that ServiceAccountNames should be empty.
> I guess I could see accepting a * as a "special" SA name if you wanted.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8941/files/2f4122d52c48cf4470e694b58842b8c9e889d2c5#r64778437>
>